lib/base/connection.js

```diff
@@ -0,0 +1,556 @@
+  }
+
+  get inTransaction() {
+    return this._activeInstance instanceof shared.driver.Transaction;
+  }
+
+  /**
+   * Creates a new connection pool with one active connection. This one initial connection serves as a probe to find out whether the configuration is valid.
+   *
+   * @param {basicCallback} [callback] A callback which is called after connection has established, or an error has occurred. If omited, method returns Promise.
+   * @return {ConnectionPool|Promise}
+   */
+
+  async open() {
+    if (this._connecting) {
+      throw new ConnectionError('Connection is connecting.')
+    }
+    if (this.opened) {
+      throw new ConnectionError('Connection is already opened.')
+    }
+
+    this._connecting = true;
+    try {
+      if (this._pool) {
+        this._connection = await this._pool.acquire(this);
+      } else {
+        this._connection = await this._connect();
+      }
+    } finally {
+      this._connecting = false;
+    }
+  }
+
+  async close() {
+    if (!this.opened) {
+      throw new ConnectionError('The connection has not been opened.')
+    }
+    if (this._closing) {
+      throw new ConnectionError('The connection is closing.')
+    }
+    this._closing = true;
+    try {
+      if (this._pool) {
+        this._pool.release(this._connection)
+      } else {
+        await this._disconnect(this._connection);
+      }
+    } finally {
+      this._closing = false;
+    }
+  }
+
+  /**
+   * if use config, open a new connection.
+   */
+  _connect() {
+    throw new Error('Not implementation.')
+  }
+
+  /**
+   * if use config, close the connection.
+   */
+  _disconnect(_tedious) {
+    throw new Error('Not implementation.')
+  }
+
+
+  /**
+   * Acquire connection from this connection pool.
+   *
+   * @param {ConnectionPool|Transaction|PreparedStatement} request Requester.
+   * @param {acquireCallback} [callback] A callback which is called after connection has been acquired, or an error has occurred. If omited, method returns Promise.
+   * @return {ConnectionPool|Promise}
+   */
+
+  acquire(request, callback) {
+    const retrn = (err) => {
+      if (typeof callback === 'function') {
+        if (err) {
+          this.emit('error', err)
+          return callback(err)
+        }
+        this._activeInstance = request;
+        return callback(null, this._connection, this.config)
+      }
+      this._activeInstance = request;
+      return shared.Promise.resolve(this._connection)
+    }
+    if (this._aborted) {
+      return retrn(new TransactionError("The transaction is automatically rolled back due to a error, pls use `rollback()` to cancel transaction.", 'EABORT'))
+    }
+    if (this._activeInstance) {
+      return retrn(new ConnectionError("Can't acquire connection for the request. There is another request in progress.", 'EREQINPROG'))
+    }
+    if (!this.opened) {
+      return retrn(new ConnectionError('The connection has not been opened.'))
+    }
+    return retrn()
+  }
```

Comment:
I think this logic really shows how fragile using a single connection is going to be, if devs attempt to make multiple queries at once against a connection they're going to be hit with errors. We need to think carefully about this.

I'm not sure why there even needs to be an `acquire` method at all because the connection isn't being released from a pool. If you instantiate the connection it is acquired by it's very nature...